File: Python/Largest-Rectangle-in-Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        seq = []
        cnt = 0
        height.append(0)
        i = 0
        while i < len(height):
            if len(seq) == 0 or height[i] >= height[seq[-1]]:
                seq.append(i)
                i += 1
            else:
                t = seq.pop()
                if len(seq) == 0:
                    cnt = max(cnt, height[t] * i)
                else:
                    cnt = max(cnt, height[t] * (i - seq[-1] - 1))
        return cnt

File: C++/Largest-Rectangle-in-Histogram.cpp
class Solution {
public:
    /**
     * @param height: A list of integer
     * @return: The area of largest rectangle in the histogram
     */
    int largestRectangleArea(vector<int> &height) {
        // write your code here
        height.push_back(0);
        stack<int> stk;
        int i = 0;
        int maxArea = 0;
        while (i < height.size()) {
            if (stk.empty() || height[stk.top()] <= height[i]) {
                stk.push(i++);
            } else {
                int t = stk.top();
                stk.pop();
                maxArea = max(maxArea, height[t] * (stk.empty() ? i : i - stk.top() - 1));
            }
        }
        return maxArea;
    }
};

